File: typescript/agentkit/src/action-providers/cdp/schemas.ts
import { z } from "zod";
import { SolidityVersions } from "./constants";

/**
 * Input schema for address reputation check.
 */
export const AddressReputationSchema = z
  .object({
    address: z
      .string()
      .regex(/^0x[a-fA-F0-9]{40}$/, "Invalid Ethereum address format")
      .describe("The Ethereum address to check"),
    network: z.string().describe("The network to check the address on"),
  })
  .strip()
  .describe("Input schema for address reputation check");

/**
 * Input schema for deploy contract action.
 */
export const DeployContractSchema = z
  .object({
    solidityVersion: z
      .enum(Object.keys(SolidityVersions) as [string, ...string[]])
      .describe("The solidity compiler version"),
    solidityInputJson: z.string().describe("The input json for the solidity compiler"),
    contractName: z.string().describe("The name of the contract class to be deployed"),
    constructorArgs: z
      .record(z.string(), z.any())
      .describe("The constructor arguments for the contract")
      .optional(),
  })
  .strip()
  .describe("Instructions for deploying an arbitrary contract");

/**
 * Input schema for deploy NFT action
 */
export const DeployNftSchema = z
  .object({
    name: z.string().describe("The name of the NFT collection"),
    symbol: z.string().describe("The symbol of the NFT collection"),
    baseURI: z.string().describe("The base URI for the token metadata"),
  })
  .strip()
  .describe("Instructions for deploying an NFT collection");

/**
 * Input schema for deploy token action.
 */
export const DeployTokenSchema = z
  .object({
    name: z.string().describe("The name of the token"),
    symbol: z.string().describe("The token symbol"),
    totalSupply: z.custom<bigint>().describe("The total supply of tokens to mint"),
  })
  .strip()
  .describe("Instructions for deploying a token");

/**
 * Input schema for request faucet funds action.
 */
export const RequestFaucetFundsSchema = z
  .object({
    assetId: z.string().optional().describe("The optional asset ID to request from faucet"),
  })
  .strip()
  .describe("Instructions for requesting faucet funds");

/**
 * Input schema for trade action.
 */
export const TradeSchema = z
  .object({
    amount: z.custom<bigint>().describe("The amount of the from asset to trade"),
    fromAssetId: z.string().describe("The from asset ID to trade"),
    toAssetId: z.string().describe("The to asset ID to receive from the trade"),
  })
  .strip()
  .describe("Instructions for trading assets");

File: python/coinbase-agentkit/coinbase_agentkit/action_providers/cdp/schemas.py
"""Schemas for CDP action providers."""

from typing import Any

from pydantic import BaseModel, Field


class AddressReputationSchema(BaseModel):
    """Input argument schema for checking address reputation."""

    address: str = Field(..., description="The Ethereum address to check")
    network: str = Field(..., description="The network to check the address on")


class DeployContractSchema(BaseModel):
    """Input argument schema for deploy contract action."""

    solidity_version: str = Field(..., description="The solidity compiler version")
    solidity_input_json: str = Field(..., description="The input json for the solidity compiler")
    contract_name: str = Field(..., description="The name of the contract class to be deployed")
    constructor_args: dict[str, Any] | None = Field(
        default=None, description="The constructor arguments for the contract"
    )


class DeployNftSchema(BaseModel):
    """Input argument schema for deploy NFT action."""

    name: str = Field(
        ...,
        description="The name of the NFT (ERC-721) token collection to deploy, e.g. `Helpful Hippos`",
    )
    symbol: str = Field(
        ...,
        description="The symbol of the NFT (ERC-721) token collection to deploy, e.g. `HIPPO`",
    )
    base_uri: str = Field(
        ...,
        description="The base URI for the NFT (ERC-721) token collection's metadata, e.g. `https://www.helpfulhippos.xyz/metadata/`",
    )


class DeployTokenSchema(BaseModel):
    """Input argument schema for deploy token action."""

    name: str = Field(..., description='The name of the token (e.g., "My Token")')
    symbol: str = Field(..., description='The token symbol (e.g., "USDC", "MEME", "SYM")')
    total_supply: str = Field(
        ..., description='The total supply of tokens to mint (e.g., "1000000")'
    )


class RequestFaucetFundsSchema(BaseModel):
    """Input schema for requesting faucet funds."""

    asset_id: str | None = Field(
        None,
        description="The asset ID to request from the faucet (defaults to ETH if not specified)",
    )


class TradeSchema(BaseModel):
    """Input argument schema for trade action."""

    value: str = Field(
        ..., description="The amount of the from asset to trade (e.g. `1.5`, `0.000001`)"
    )

    from_asset_id: str = Field(
        ...,
        description="The from asset ID to trade (e.g. `eth`, `0x036CbD53842c5426634e7929541eC2318f3dCF7e`)",
    )
    to_asset_id: str = Field(
        ...,
        description="The to asset ID to receive from the trade (e.g. `eth`, `0x036CbD53842c5426634e7929541eC2318f3dCF7e`)",
    )
